### PR TITLE
ARROW-7194: [Rust] Fix CSV writer recursion issues

### PR DESCRIPTION
**Pull request by LuoZijun**
When we use both the `arrow@master` and `tokio-reactor@0.1` libraries in a project, there is no way to compile the code, even in the latest nightly version.

### CODE REVIEW

**Comment by kszucs** [comment]
@LuoZijun please describe what this change is about

**Comment by LuoZijun** [comment]
> @LuoZijun please describe what this change is about

Done.

**Comment by nevi-me** [comment]
Thanks @LuoZijun, I've seen this error before in one of my projects, but wasn't sure of what was causing it. I've opened https://issues.apache.org/jira/browse/ARROW-7194

File: rust/arrow/src/csv/writer.rs [hunk 1]
```diff
@@ -105,12 +105,12 @@ pub struct Writer<W: Write> {
...
-{
+impl<W: Write> Writer<W> {
     /// Create a new CsvWriter from a writable object, with default options
     pub fn new(writer: W) -> Self {
+        let delimiter = b',';
+        let mut builder = csv_crate::WriterBuilder::new();
+        let writer = builder.delimiter(delimiter).from_writer(writer);
         Writer {
             writer,
             delimiter: b',',
```
Comment (nevi-me) [inline rust/arrow/src/csv/writer.rs]: would it be possible to reuse the `delimiter` variable here?

File: rust/arrow/src/csv/writer.rs [hunk 2]
```diff
@@ -361,6 +359,9 @@ impl WriterBuilder {
 
     /// Create a new `Writer`
     pub fn build<W: Write>(self, writer: W) -> Writer<W> {
+        let delimiter = self.delimiter.unwrap_or(b',');
+        let mut builder = csv_crate::WriterBuilder::new();
+        let writer = builder.delimiter(delimiter).from_writer(writer);
         Writer {
             writer,
             delimiter: self.delimiter.unwrap_or(b','),
```
Comment (nevi-me) [inline rust/arrow/src/csv/writer.rs]: same re. reusing `delimiter`

File: rust/arrow/src/util/string_writer.rs [hunk 3]
```diff
@@ -63,6 +63,7 @@
 use std::cell::RefCell;
 use std::io::{Error, ErrorKind, Result, Write};
 
+#[derive(Debug)]
```
Comment (nevi-me) [inline rust/arrow/src/util/string_writer.rs]: I think you needed this in your example to reproduce the bug, but we should not need here. May you please remove the derive here?

**Review by nevi-me — APPROVED** [review]
LGTM @LuoZijun, may you please address the minor comments, then we can merge this.

If you also don't mind, may you please create a JIRA account at https://issues.apache.org/jira so we can assign the ticket/issue to your name 😃

File: rust/arrow/src/util/string_writer.rs (hunk 3 above)
Comment (LuoZijun) [reply rust/arrow/src/util/string_writer.rs]: removed.

File: rust/arrow/src/csv/writer.rs (hunk 2 above)
Comment (LuoZijun) [reply rust/arrow/src/csv/writer.rs]: done.

File: rust/arrow/src/csv/writer.rs (hunk 1 above)
Comment (LuoZijun) [reply rust/arrow/src/csv/writer.rs]: done.

**Comment by LuoZijun** [comment]
> LGTM @LuoZijun, may you please address the minor comments, then we can merge this.

done.

> If you also don't mind, may you please create a JIRA account at https://issues.apache.org/jira so we can assign the ticket/issue to your name 😃

I created a JIRA account.
The name is luozijun, the full name Luozijun.

**Comment by nevi-me** [comment]
@kszucs may you please assign https://issues.apache.org/jira/browse/ARROW-7194 to Luozijun. I don't have permission to do so

**Comment by kou** [comment]
Done.

**Pull request closed without merging**